wine: update to 10.2

fvalasiad opened the pull request:
Also add doc about the new WoW64 build option and update the alert in the script.

Broken wine64 symlink was removed.

Hotfix for updated wine-tools, the script now passes the 0th argument to wine properly(Props to @oreo639).

The removal of the wine/wine64 scheme caused issues with users not being able to install both wine and wine-32bit together. It's being addressed by now installing to `/usr/libexec/wine/wine32` and `/usr/libexec/wine/wine64` respectively.

There is a bug report that the 32bit build is broken, uncertain if it's related to the issue described above, left as a [WIP] & draft until resolved.
logs: https://gist.github.com/oreo639/cf2f5f8cb27475db87fc58051e69ddf4

fvalasiad commented:
@Hoshpak Hello!

Would be great if you could add the needs-testing tag.

Thank you very much!

fvalasiad commented:
Update: You can now install both as expected, winecfg and other tools work properly, but the failure to load kernel32.dll remains. Will keep investigating.

summrum commented:
Thank you for notifying about the updated package pull request in the [32/64 bit incompatibility issue](https://github.com/void-linux/void-packages/issues/54456). I have built the native x86_64 and i686 (i.e. ```xbps-src -A i686 pkg wine```). The updated wine-10.2_2, wine-common-10.2_2 and wine-32bit-10.2_2 all install fine but I have issues running 32 bit executables with wine, for example:
```
wine: failed to load /usr/lib32/wine/i386-windows/ntdll.dll error 4000000e
0024:err:environ:run_wineboot failed to start wineboot 1
wine: failed to load /usr/lib32/wine/i386-windows/ntdll.dll error 4000000e
0024:err:win:get_desktop_window failed to start explorer 1
0038:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0038:err:winediag:nodrv_CreateWindow L"The explorer process failed to start."
0024:err:dbghelp:pe_load_dbg_file Couldn't find .DBG file "msvcrt.dbg" (L"")
```
I am unsure whether I may have made a mistake with compiling the 32bit version for my machine, so if there is something I may have missed please let me know. Thanks again.

fvalasiad commented:
Update: further inspection has me assuming that the prefix initialization step is the one malfunctioning.

The error as everyone has seen is complaining about missing dlls, those dlls are meant to be placed inside the respective WINEPREFIX.

When we run /usr/libexec/wine/wine64, the targeted prefix is populated properly, and the process runs as expected.

/usr/libexec/wine/wine32 though, while claiming that it properly initialized the prefix, leaves behind an empty prefix, and when it tries to run anything, it complains cause it cannot find the dlls in there.

fvalasiad commented:
Update 2: Thank god for strace.

Check this out:
```
[pid  4735] read(27, "\0\0\0\0\0\0\0\0\0\0\0\0H\0\0\0\\\0?\0?\0\\\0C\0:\0\\\0w\0"..., 131) = 131                      [pid  4735] openat(AT_FDCWD, "/usr/lib32/wine/x86_64-windows/wineboot.exe", O_RDONLY|O_NONBLOCK) = -1 ENOENT (No such file or directory)                                                                                                    [pid  4735] write(28, "\17\0\0\300\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 64) = 64
```

As to why it tries to access x86_64-windows instead of i386-windows, here is what I found after examining the strace log:

where did pid 4735 come from?

From a clone from process 4734!

Where did pid 4734 come from?

from a clone from process 4730!

What is 4730? A process id, and if we look at one of its threads, 4734, we see the following:
```
[pid  4734] rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0                                                              [pid  4734] execve("/usr/bin/wineserver", ["/usr/bin/wineserver"], 0xff982cd0 /* 24 vars */ <unfinished ...>
[pid  4730] <... clone3 resumed>)       = 4734                                                                        [pid  4730] munmap(0xf7f13000, 36864)   = 0
[pid  4734] <... execve resumed>)       = 0
[pid  4730] rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0                                                              [pid  4734] [ Process PID=4734 runs in 64 bit mode. ] 
```

let's isolate the 4734 logs and look further down, we find:

```
[pid  4734] read(10, "#!/bin/sh -e\n# Based on the Debi"..., 8192) = 516
[pid  4734] faccessat2(AT_FDCWD, "/usr/libexec/wine/wineserver64", X_OK, AT_EACCESS) = 0
[pid  4734] execve("/usr/libexec/wine/wineserver64", ["/usr/libexec/wine/wineserver64", "-p0"], 0x5588205c4828 /* 24 v
``

So there you go, the 32bit wine executable launches the 64bit wineserver.
Is this the reason why it doesn't work? I don't know! Just an assumption.

onlylunix commented:
> please test the patch so we can get it merged
I assembled, installed and got an error
`0 $ xbps-query -s wine
[*] wine-10.2_2           Run Microsoft Windows applications
[*] wine-common-10.2_2    Run Microsoft Windows applications - common files
[*] wine-mono-9.4.0_1     Mono built for running .NET applications with WINE
[*] wine-tools-10.2_2     Run Microsoft Windows applications - development tools
[*] winetricks-20240105_1 Script to install various redistributable runtime libraries in Wine
0 $ winecfg
it looks like wine-32bit is missing, you should install it.
the multilib repository needs to be enabled first.  as root, please
execute "xbps-install -S void-repo-multilib && xbps-install -S wine-32bit"
/usr/bin/wineserver: 11: test: ELF: unexpected operator
/usr/bin/wineserver: 18: test: ELF: unexpected operator
error: unable to determine architecture.
`

fvalasiad commented:
AAAAA I forgot to put the angle brackets! It's because in the script I use `$arch` instead of `"$arch"`

I forgot to put it in the PR, sorry, I am updating it right now.

To avoid recompiling, you can edit "/usr/bin/wineserver" with your editor of choice and add them yourself. Sorry!

onlylunix commented:
>  edit "/usr/bin/wineserver" 
after editing:

0 $ winecfg
it looks like wine-32bit is missing, you should install it.
the multilib repository needs to be enabled first.  as root, please
execute "xbps-install -S void-repo-multilib && xbps-install -S wine-32bit"
002c:err:winediag:getaddrinfo Failed to resolve your host name IP
0144:err:environ:init_peb starting L"C:\\windows\\syswow64\\rundll32.exe" in experimental wow64 mode
wine: failed to load L"\\??\\C:\\windows\\syswow64\\ntdll.dll" error c0000135
0 $ 

after editing
messages are alarming
i am testing

fvalasiad commented:
> > edit "/usr/bin/wineserver"
> 
> after editing: 0 $ winecfg it looks like wine-32bit is missing, you should install it. the multilib repository needs to be enabled first. as root, please execute "xbps-install -S void-repo-multilib && xbps-install -S wine-32bit" 002c:err:winediag:getaddrinfo Failed to resolve your host name IP 0144:err:environ:init_peb starting L"C:\windows\syswow64\rundll32.exe" in experimental wow64 mode wine: failed to load L"\??\C:\windows\syswow64\ntdll.dll" error c0000135 0 $
> 
> messages are alarming i am testing

Ah you enabled WoW on x86_64? I haven't tested that one, I tested on musl though and was able to run a 32bit app, obviously with a few issues given its experimental nature.

The 32bit binary warning is present when you use wine without having wine-32bit. You don't need it if you use WoW, but it's there nevertheless and should be ignored. Perhaps we can somehow test and silence the warning if we detect the option enabled in the future.

onlylunix commented:
> Ah you enabled WoW on x86_64?

how to do it?

fvalasiad commented:
You've already done it, as suggested by `0144:err:environ:init_peb starting L"C:\windows\syswow64\rundll32.exe" in experimental wow64 mode`

Unless wine handles it automatically if it doesn't find a 32bit wine binary? Uncertain! But yeah if you aren't interested in also participating in the beta of the new WoW, you need to also compile for 32bit and install that as well, for games and stuff!
```
./xbps-src -A i686 binary-bootstrap
./xbps-src -A i686 pkg wine
```

Pull my update before running the build!

onlylunix commented:
So you can't do without wine-32bit?

fvalasiad commented:
Shortly,

When microsoft transitioned to 64bit, they had a giant collection of legacy 32bit apps that they couldn't let go of.

They implemented WoW64, which would allow you to run 32bit apps on 64bit installs.

Many programs today aren't fully 64bit, including games.

Wine for all this time would need both a 64bit & 32bit install to imitate WoW64 properly. As of version 9.0, they introduced a new experimental WoW64 mode that allows a 64bit build to run 32bit builds on its own, removing the requirement for a 32bit wine install.

This though is experimental, and currently we have it enabled just for Musl, since musl doesn't have a 32bit version anyways so there is no tradeoff to be made.

Hoshpak commented:
@fvalasiad Something still isn't right, I'm afraid. I built both wine and wine-32bit, installed them and when I try to launch an application, it will create a wineprefix, but after that, it fails:

`002c:err:winediag:getaddrinfo Failed to resolve your host name IP
012c:err:environ:init_peb starting L"Z:\\tmp\\7z2409-x64.exe" in experimental wow64 mode
wine: failed to load L"\\??\\C:\\windows\\syswow64\\ntdll.dll" error c0000135
Application could not be started, or no application associated with the specified file.
ShellExecuteEx failed: 0024:err:start:fatal_error FormatMessage failed
`

It looks like the binary is now unable to detect the presence of the 32 bit installation and automatically selects the experimental wow64 mode.

Creating a 32bit-only wineprefix and launching 32 bit executables by calling the wine32 binary directly works, but not by using the launcher script.

fvalasiad commented:
Yeah that aligns with the report from onlylunix, I would not say it uses the experimental wow64 mode, as I tested that on musl and it works fine. More like it tries to use it but fails once it cannot find syswow64/ntdll.dll which I assume is only present if you supply the configure options to include the experimental wow64 support.

Truth is I am using musl so these issues flew right over my head. Guess I will spin up a VM and try to launch a game from now on with each update.

Ah, what a bummer, more digging it is. Once again it's 21:18 here so not the greatest time to be debugging, and unfortunately tomorrow is gonna be a damn busy day so I ain't gonna be active, I absolutely invite anyone interested to join me on the bughunt though!

If I were to tackle this, I'd probably once again use strace and look for failed `open(2)` syscalls, safe to assume that it tries to open some paths and after it fails to find anything it spits that report. This is only for someone who doesn't know the wine codebase though!

summrum commented:
I just came to report the same; I have to create a new 32 bit only wineprefix (which I did not previously have to do with 10.1) to run 32 bit applications. It looks to me like with the wrapper it is trying to launch in wow64 mode, but this is not enabled for glibc build (only musl) and I built for glibc without adding the wow64 build option.
```
0140:err:environ:init_peb starting L"Z:\\home\\summrum\\.wine\\drive_c\\32bitexecutable.exe" in experimental wow64 mode
wine: failed to load L"\\??\\C:\\windows\\syswow64\\ntdll.dll" error c0000135
```

Hoshpak commented:
@fvalasiad I think I found something. The original wrapper script used the 32bit binary if available and used that to leverage the wow64 functionality. The changes in this PR turn this logic in the opposite direction, launching the 64bit binary if available and expect it to be able to handle the wow64 stuff which seems to lead to it trying the experimental mode.

I followed your advice of using strace to get closer to the issue and found that when using the wine 32 bit binaries (both main binary and wineserver), it looks for files in /usr/lib32/wine/x86_64-unix which isn't there (the 64 bit build is installed in /usr/lib/wine/x86_64-unix). My next attempt was to just copy x86_64-unix and x86_64-windows over to /usr/lib32/wine and voila, I could execute binaries and create wine prefixes again.

Of course this is not a proper solution for the issue yet but I hope this helps in finding one. It's getting rather late here too so I'm not able to spend any more time on this for today and won't be back until tomorrow in the evening but since the package has been reverted to a working release already, we can take the time we need here to find a proper solution.

fvalasiad commented:
@Hoshpak Thanks for the help, I actually went and used strace too and I found it trying to access /usr/lib/wine/i386-unix/wine, which it cannot, since the real location is /usr/lib32/wine/i386-unix

I followed your approach by simlinking the directory there, and voila, I can also launch apps!

As far as the "reverse" approach is concerned, yeah I flipped em on purpose since I noticed it was creating 32bit prefixes, I assumed that was their intention anyways.

About a proper solution, building wine doc mentions that if you want to use wow you need to supply the --enable-win64 configure option, which we do, but then you also need to point the 32bit build to the 64bit build using `--with-wine64=<wine64 build dir>`

I assume we are cheating this somehow by placing the wine64 build files to some location which wine was finding up until now. I am thinking maybe if we can leverage that option to guide wine to find the 32bit install....

fvalasiad commented:
@Hoshpak Oh hey I just noticed you used the 32bit binaries, I used the 64bit ones, explains the lookup at lib vs lib32?

Yeah confirmed that in the million of lines of strace log, not a single instance of lib32 is found. So wine actually expects i386-* and x86_64-* to be in the same location, either under /usr/lib or /usr/lib32 given some condition.

we could actually go and symlink those together but yeah this doesn't qualify as a proper solution =')

fvalasiad commented:
So reading the description from `configure.ac`:
```
AC_ARG_WITH(system-dllpath,AS_HELP_STRING([--with-system-dllpath=PATH],[load external PE dependencies from colon-separated path PATH]),
            AC_SUBST(system_dllpath,[$withval]))
AC_ARG_WITH(wine-tools,AS_HELP_STRING([--with-wine-tools=DIR],[use Wine tools from directory DIR]))
AC_ARG_WITH(wine64,    AS_HELP_STRING([--with-wine64=DIR],[use the 64-bit Wine in DIR for a Wow64 build]))
```

Adding `--with-wine64=/usr/lib/wine` will work or am I too optimistic? What about `--with-wine64=/usr/lib/wine/x86_64-unix`?

Can't know unless we try huh?

fvalasiad commented:
@Hoshpak I am kinda starting to doubt the usefulness of the wrapper scripts? Help me on that.

The first thing that struck me as odd was the checks for `wineserver` from `strace(1)`, notice the first location that 32bit wine looked for wineserver:
`[pid  4731] execve("/usr/lib32/wine/../../libexec/wine/wineserver", ["/usr/lib32/wine/../../libexec/wi"...], 0xff982cd 0 /* 24 vars */) = -1 ENOENT (No such file or directory)`

Also notice, how wine further down the line seemingly prefers to use the binaries under `/x86_64-unix/` and `/i686-unix/` respectively instead of relying on `/usr/bin` for anything, and it makes sense because that's how it differentiates `x86_64` from `i686`.

Additionally you mentioned that you tested the symlink idea, but you did not specify, you symlinked the 32bit lib wine dirs to 64bit respectively, but did  you do the same vice versa? I've concluded that WoW64 works either way, either by picking the 64bit wine or 32bit, as long as it finds the desired `i686-*` and `x86_64-*` dirs under `/usr/lib/wine` and `/usr/lib32/wine` respectively.

Additionally I can see why you would not want broken symlinks to be installed nevertheless when you install either package, I am just curious is this a practical reason or design decision? I don't see how placeholder symlinks hurt anyone after all!

Honestly at this point it feels like we are trying to fix a "hacky" solution, so once I am back home this afternoon I will attempt a redesign. We can keep the wrapper scripts per-se but not have the build modified at all by specifying `--bindir=/usr/libexec/wine`

I will test and report back, feel free to stop me in advance if you think this is a bad idea!

Hoshpak commented:
@fvalasiad I just had the build process link the 32 bit directories to /usr/lib which has the benefit of the symlinks being in the wine-32bit package (and not the regular i686 package) so they are removed when the package is uninstalled. Doing it the other way would also work but would make the logic more complex since we would need to differentiate if 32 bit and 64 bit package are installed and clean up afterwards, which would make it harder to implement it properly.

I guess you could call it a design decision to not have dangling symlinks in the system. From my point of view, the purpose of a package manager is to allow for the clean installation and uninstallation of packages and should not leave any files behin that were part of the installation in system directories in case the package is uninstalled. Creating symlinks at runtime or creating dangling ones that might be needed when another package is installed but not when it's not, would violate that principle.

The wrapper scripts could probably be somewhat simplified but unless we find another solution, they would still serve the purpose of making it possible to install 32 bit only, 64 bit only or the combination of both on glibc systems and being able to just call the wine command to execute binaries for the architecture you installed wine on your system.

 I'm not sure I understand the idea behind setting the bindir explicitely but if you think we could simplify our setup this way, it's for sure worth a try.

fvalasiad commented:
@Hoshpak Ah I wasn't looking into creating symlinks at runtime, I was thinking of adding something at `post_install()`, like hey add a symlink to `/usr/lib32/wine/x86_64-*` from `/usr/lib/wine/` when you install the 64bit wine package and do the same backwards when you install the 32bit package, also removing them respectively when you uninstall each package.

The comment about `--libdir` was that seeing the strace log of running the 32bit wine executable, it seemed as if `wine` knew what it was doing by respectively trying to access `/usr/lib32/wine/x86_64-*` and `/usr/lib32/wine/i686-*` depending on what it was looking for.

It was only a fallback to access the wineserver by `exec(2)`-ing `/usr/bin/wineserver`. I am thus suggesting that maybe by specifying `--bindir=/usr/libexec/wine` we somehow "break" wine's working config which then forces us to deal with the problem ourselves in our wrapper script.

So, I am suggesting, perhaps instead of going forward with trying to make the wrapper script work, we instead focused on making the environment more predictable for wine, and let it work for us.

The way I understand it after all, is that the purpose of the wrapper script is limited to wine's user interface, and should not intermix with wine`s practical internals, which it currently does, as indicated by it trying to access `/usr/lib32/wine/../../libexec/wine/wineserver`

fvalasiad commented:
So here are some results after removing `--bindir` and anything surrounding `/usr/libexec`

1. Wine successfully manages to find both its 32bit & 64bit variant for WoW64 by launching either executable of the two.

2. Additionally it creates the symlinks itself back & forth.

3. Removing say 64bit wine, leaves the dangling symlinks created by wine behind. Same goes for 32bit

4. In my current template, even after the removal of 64bit wine, if 32bit wine and wine-common remains, the 64bit ELFs at /usr/bin/ remain, and they are broken for usage. This is probably a template issue which could be dealt with.

I absolutely agree about the purpose of a package manager, and I find leaving files behind after  an uninstall undesirable if not unacceptable.

Nevertheless though it seems more promising as an approach, since it works without our interference, so I am gonna spend some extra time here to see if I can get it fully working. Tomorrow is once again gonna be a really busy days, but I will try my best to allocate some time. Bear with me :)

onlylunix commented:
Bro, we don't have enough teachers in the northern camps, you're a good candidate if you can't handle this problem. Joke

fvalasiad commented:
@Hoshpak this is the idea, polished, this setup should be fully working in any config.

It relies on INSTALL and REMOVE scripts which make sure to create and remove the symbolic links if both 64 and 32bit installs are detected and if any of those is removed respectively.

This makes it so that there are no dangling symbolic links at any point in time, additionally WoW64 is enabled regardless of which one you call, wine/wine32/wine64.

Review, test, and ping me back!

Cheers!

@onlylunix Uncertain if you are referring to me but I am still chasing after my Bachelor so that would be hard = D.

fvalasiad commented:
FYI `lib32symlinks` should also work properly with this, the only difference would be that you would need to launch the 64bit executable exclusively to have WoW64, aka `wine32` would be for 32bit windows apps exclusively.

Hoshpak commented:
I just gave it a test and it seems to work fine for new installations of the packages, installation order doesn't matter, no unnecessary symlinks created, no dangling ones left behind. However my first attempt was to update from already installed wine 10.1 packages and that didn't create the symlinks resulting in a nonfunctional installation since the update path is explicitely excluded so that is something that still needs to be solved. I'm not sure what the best solution here is. We could use "ln -sf" and remove the update guard which would make the symlinkling an idempotent operation (given that no directory already exists at the target path which should be the case). Another option would be to check each target path explicitely and create the symlink only if necessary.

The removal of a package currently produces some error messages about nonexisting paths if the symlinks weren't created by a previous installation but that's a cosmetic issue and not a functional one.

fvalasiad commented:
Solid idea! Removed the update guard and added `-f` to each `-ln` call. Additionally I also added `2>dev/null` to all `unlink` calls in `REMOVE` to improve cosmetics.

Quite honestly the REMOVE script could even remove the `if` statement altogether, just nuke all symlinks if wine is removed, if both 32 & 64bit are installed the script will behave as expected, if only a single one is installed the script will just redirect the errors to `/dev/null` without any side effects whatsoever.
